### PR TITLE
[SPARK-41509][SQL] Only execute `Murmur3Hash` on aggregate expressions for semi-join runtime filter.

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Currently, Spark runtime filter supports bloom filter and in subquery filter.
The in subquery filter always execute `Murmur3Hash` before aggregate the join key.

Because the data size before aggregate lager than after, we can only execute `Murmur3Hash` for aggregation and it will reduce the number of calls to `Murmur3Hash` and improve performance.

### Why are the changes needed?
Improve performance for semi-join runtime filter.

### Does this PR introduce _any_ user-facing change?
'No'.
Just update the inner implementation.

### How was this patch tested?
Manually test TPC-DS.
```
spark.sql.optimizer.runtime.bloomFilter.enabled=false
spark.sql.optimizer.runtimeFilter.semiJoinReduction.enabled=true
```

TPC-DS data size: 2TB.
This improvement is valid for below TPC-DS test cases and no regression for other test cases.

| TPC-DS Query   | Before(Seconds)  | After(Seconds)  | Speedup(Percent)  |
|  ----  | ----  | ----  | ----  |
| q23a | 542.03 | 539.06 | 0.55% |
| q23b | 776.7 | 769.74 | 0.90% |
| q24a | 442.25 | 436.75 | 1.26% |
| q24b | 436.16 | 432.86 | 0.76% |
| q50 | 200.92 | 193.36 | 3.91% |
| q64 | 426.73 | 421.03 | 1.35% |
| q67 | 987.79 | 956.82 | 3.24% |
| q93 | 397.26 | 393.9 | 0.85% |
| All TPC-DS | 8392.49 | 8302.56 | 1.08% |

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala [hunk 1]
```diff
@@ -99,16 +99,17 @@ object InjectRuntimeFilter extends Rule[LogicalPlan] with PredicateHelper with J
       filterCreationSideExp: Expression,
       filterCreationSidePlan: LogicalPlan): LogicalPlan = {
     require(filterApplicationSideExp.dataType == filterCreationSideExp.dataType)
-    val actualFilterKeyExpr = mayWrapWithHash(filterCreationSideExp)
-    val alias = Alias(actualFilterKeyExpr, actualFilterKeyExpr.toString)()
+    val alias = Alias(filterCreationSideExp, filterCreationSideExp.toString)()
     val aggregate = ColumnPruning(Aggregate(Seq(alias), Seq(alias), filterCreationSidePlan))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala:103]: shall we simply add the hash expression for the second `Seq(alias)`?

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala:103]: This will increase the calculation.

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala:103]: I have ran the TPC-DS with you suggestion, regression appears.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala:103]: I don't get it. How is `Project(expr, Agg(...))` different from `Agg(grouping, expr)`? `Agg` has a project list and can do the job of `Project` as well.

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala:103]: I'm sorry. I forgot `spark.sql.optimizer.runtime.bloomFilter.enabled=false`.

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala:103]: I will test again.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala [hunk 2]
```diff
@@ -101,7 +101,8 @@ object InjectRuntimeFilter extends Rule[LogicalPlan] with PredicateHelper with J
     require(filterApplicationSideExp.dataType == filterCreationSideExp.dataType)
     val actualFilterKeyExpr = mayWrapWithHash(filterCreationSideExp)
     val alias = Alias(actualFilterKeyExpr, actualFilterKeyExpr.toString)()
-    val aggregate = ColumnPruning(Aggregate(Seq(alias), Seq(alias), filterCreationSidePlan))
+    val aggregate =
+      ColumnPruning(Aggregate(Seq(filterCreationSideExp), Seq(alias), filterCreationSidePlan))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala:105]: you changed the group by expr, why?

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
is the benchmark result update to update with this new change?

**Comment by beliefer** [comment]
After my test, the results are similar to those of the previous method.

**Comment by beliefer** [comment]
> is the benchmark result update to update with this new change?

After my test, the results are similar to those of the previous method.

**Comment by beliefer** [comment]
@cloud-fan Thank you!

**Pull request closed without merging**